mlflow/projects/__init__.py

```diff
@@ -406,31 +356,32 @@ def _run_project(project, entry_point, work_dir, parameters, use_conda, storage_
         commands.append("source activate %s" % _get_conda_env_name(conda_env_path))
 
     # Create a new run and log every provided parameter into it.
-    active_run = tracking.start_run(experiment_id=experiment_id,
-                                    source_name=project.uri,
-                                    source_version=tracking._get_git_commit(work_dir),
-                                    entry_point_name=entry_point,
-                                    source_type=SourceType.PROJECT)
+    exp_id_for_run = experiment_id or tracking._get_experiment_id()
+    assert tracking._active_run is None
+    active_run = tracking._create_run(
+        experiment_id=exp_id_for_run, source_name=project.uri,
+        source_version=tracking._get_git_commit(work_dir), entry_point_name=entry_point,
+        source_type=SourceType.PROJECT)
+    assert tracking._active_run is None
     if parameters is not None:
         for key, value in parameters.items():
             active_run.log_param(Param(key, value))
     # Add the run id into a magic environment variable that the subprocess will read,
     # causing it to reuse the run.
     exp_id = experiment_id or tracking._get_experiment_id()
     env_map = {
-        tracking._RUN_NAME_ENV_VAR: active_run.run_info.run_uuid,
+        tracking._RUN_ID_ENV_VAR: active_run.run_info.run_uuid,
         tracking._TRACKING_URI_ENV_VAR: tracking.get_tracking_uri(),
         tracking._EXPERIMENT_ID_ENV_VAR: str(exp_id),
     }
 
     commands.append(run_project_command)
     command = " && ".join(commands)
     eprint("=== Running command: %s ===" % command)
-    try:
-        process.exec_cmd([os.environ.get("SHELL", "bash"), "-c", command], cwd=work_dir,
-                         stream_output=True, env=env_map)
-        tracking.end_run()
-        eprint("=== Run succeeded ===")
-    except process.ShellCommandException:
-        tracking.end_run("FAILED")
-        eprint("=== Run failed ===")
+
+    if block:
+        _launch_local_command(active_run, command, work_dir, env_map)
+    else:
+        multiprocessing.Process(
```

Comment:
In asynchronous/non-blocking mode, we launch a subprocess that monitors the run & reports its status to the tracking server once it terminates. This avoids the scenario e.g. where:

* The current process executes `mlflow.projects.run(async=True)` and continues 
* The current process does something to cause an exception (e.g. 1/0) and exits
* The subprocess that runs `command` continues to run, calls `log_metric` etc and exits, but never calls `end_run` & so the run is never recorded as finished

Reply:
Have you double-checked that this way of launching a process lets it continue running even if the current process crashes? Is it documented somewhere?